### PR TITLE
Completion block fix

### PR DESCRIPTION
**Pull request by carlbrown**
Fix for completion block issue #36.  

Also includes a fix for #48 that caused me to retract my last attempt (PR #47).

Also adds at least one test for #46 while I was at it (but more tests would still be useful there).

### CODE REVIEW

File: Sources/HTTP/HTTPStreamingParser.swift [hunk 1]
```diff
@@ -284,21 +304,29 @@ public class StreamingParser: HTTPResponseWriter {
     func bodyReceived(data: UnsafePointer<Int8>?, length: Int) -> Int32 {
         processCurrentCallback(.bodyReceived)
         guard let data = data else { return 0 }
+        if shouldStopProcessingBody {
```
Comment (seabaylea) [inline Sources/HTTP/HTTPStreamingParser.swift:307]: Optimally, this should be before the assignment.

File: Sources/HTTP/HTTPStreamingParser.swift [hunk 2]
```diff
@@ -284,21 +304,29 @@ public class StreamingParser: HTTPResponseWriter {
...
-                            finished = true
-                        }), &stop)
+                        //OK, this sucks.  We can't access the value of the `inout` inside this block
+                        //  due to exclusivity. Which means that if we were to pass a local variable, we'd
+                        //  have to put a semaphore or something up here to wait for the block to be done before
+                        //  we could get its value and pass that on to the instance variable. So instead, we're
+                        //  just passing in a pointer to the internal ivar. But that ivar can't be modified in
+                        //  more than one place, so we have to put a semaphore around it to prevent that.
+                        _shouldStopProcessingBodyLock.wait()
+                        handler(.chunk(data: chunk, finishedProcessing: {self._shouldStopProcessingBodyLock.signal()}), &_shouldStopProcessingBody)
```
Comment (seabaylea) [inline Sources/HTTP/HTTPStreamingParser.swift:327]: `&_shouldStopProcessingBody` isn't protected by the lock....

Comment (seabaylea) [reply Sources/HTTP/HTTPStreamingParser.swift:327]: Do we actually need to wait for the async completion handler to run before we can set stop? Don't we want to set it immediately (sync) even if an async `writeBody()` is still going to happen?

**Comment by seabaylea** [comment]
Out of interest, is there any reason why stop would be a flag rather than an API call on the HTTPResponseWriter, i.e. alongside done() and abort()?

**Comment by carlbrown** [comment]
>Out of interest, is there any reason why stop would be a flag rather than an API call on the HTTPResponseWriter, i.e. alongside done() and abort()?

It's not my API, I just implemented it.

File: Sources/HTTP/HTTPStreamingParser.swift (hunk 1 above)
Comment (carlbrown) [reply Sources/HTTP/HTTPStreamingParser.swift:307]: Locking the semaphore to check the Bool value should take precedence over the nil-check?

File: Sources/HTTP/HTTPStreamingParser.swift (hunk 2 above)
Comment (carlbrown) [reply Sources/HTTP/HTTPStreamingParser.swift:327]: > &_shouldStopProcessingBody isn't protected by the lock....

?
That's what line 326 does.

> Do we actually need to wait for the async completion handler to run before we can set stop? Don't we want to set it immediately (sync) even if an async writeBody() is still going to happen

We can't set `stop` here - it's passed as in `inout` to the API, and it's set by the `handler`.

Comment (seabaylea) [reply Sources/HTTP/HTTPStreamingParser.swift:327]: The completion handler that releases the lock runs async (separate thread) - there's no guarantee that its going to call signal() after `_shouldStopProcessingBody` is modified, so its effectively unprotected.

I was wondering whether your intent was to block access to `_shouldStopProcessingBody` until after the completion handler has run.

File: Sources/HTTP/HTTPStreamingParser.swift (hunk 1 above)
Comment (seabaylea) [reply Sources/HTTP/HTTPStreamingParser.swift:307]: Sorry, I'm getting ahead of myself. I'm not convinced we need any lock protection around a stop flag. We're dealing with two serial queues - one reader and one writer. The only logic that runs on the writer queue is the writeXXXX() calls, so if reads and updates to the stop flag are only by code executing on the reader queue then we don't need a lock.

If we don't need to lock it, it should optimally be before the nil check and assign (unless data = nil is a more common event?)

**Review by ianpartridge — COMMENTED** [review]
Please break this down into one PR per problem, or a series of commits that build on each other and can be reviewed independently.

**Comment by carlbrown** [comment]
Removing in favor of #51 (and the several others that will be stacked on top of it)

**Pull request closed without merging**